data/src/test/java/org/apache/iceberg/io/TestDVWriters.java

```diff
@@ -100,6 +114,211 @@ public void testBasicDVs() throws IOException {
+    assertThat(table.currentSnapshot().addedDeleteFiles(table.io())).hasSize(1);
+    assertThat(table.currentSnapshot().removedDeleteFiles(table.io())).hasSize(1);
+
+    // verify correctness
+    assertRows(ImmutableList.of(toRow(3, "aaa")));
+  }
+
+  @TestTemplate
+  public void testApplyPartitionScopedPositionDeletes() throws IOException {
+    assumeThat(formatVersion).isEqualTo(2);
+
+    FileWriterFactory<T> writerFactory = newWriterFactory(table.schema());
+
+    // add the first data file with 3 records
+    List<T> rows1 = ImmutableList.of(toRow(1, "aaa"), toRow(2, "aaa"), toRow(3, "aaa"));
+    DataFile dataFile1 = writeData(writerFactory, parquetFileFactory, rows1, table.spec(), null);
+    table.newFastAppend().appendFile(dataFile1).commit();
+
+    // add the second data file with 3 records
+    List<T> rows2 = ImmutableList.of(toRow(4, "aaa"), toRow(5, "aaa"), toRow(6, "aaa"));
+    DataFile dataFile2 = writeData(writerFactory, parquetFileFactory, rows2, table.spec(), null);
+    table.newFastAppend().appendFile(dataFile2).commit();
+
+    // add a position delete file with deletes for both data files
+    DeleteFile deleteFile =
+        writePositionDeletes(
+            writerFactory,
+            ImmutableList.of(
+                Pair.of(dataFile1.location(), 0L),
+                Pair.of(dataFile1.location(), 1L),
+                Pair.of(dataFile2.location(), 0L)));
+    table.newRowDelta().addDeletes(deleteFile).commit();
+
+    // verify correctness with the position delete file
+    assertRows(ImmutableList.of(toRow(3, "aaa"), toRow(5, "aaa"), toRow(6, "aaa")));
+
+    // upgrade the table to V3 to enable DVs
+    table.updateProperties().set(TableProperties.FORMAT_VERSION, "3").commit();
+
+    // write a DV, applying old positions but keeping the position delete file in place
+    DVFileWriter dvWriter =
+        new BaseDVFileWriter(
+            fileFactory,
+            new PreviousDeleteLoader(table, ImmutableMap.of(dataFile2.location(), deleteFile)));
+    dvWriter.delete(dataFile2.location(), 1L, table.spec(), null);
+    dvWriter.close();
+
+    // validate the writer result
+    DeleteWriteResult result = dvWriter.result();
+    assertThat(result.deleteFiles()).hasSize(1);
+    assertThat(result.referencedDataFiles()).containsOnly(dataFile2.location());
+    assertThat(result.referencesDataFiles()).isTrue();
+    assertThat(result.rewrittenDeleteFiles()).isEmpty();
+    DeleteFile dv = Iterables.getOnlyElement(result.deleteFiles());
+
+    // commit the DV, ensuring the position delete file remains
+    commit(result);
+    assertThat(table.currentSnapshot().addedDeleteFiles(table.io())).hasSize(1);
+    assertThat(table.currentSnapshot().removedDeleteFiles(table.io())).isEmpty();
+
+    // verify correctness with DVs and position delete files
+    assertRows(ImmutableList.of(toRow(3, "aaa"), toRow(6, "aaa")));
+
+    // verify the position delete file applies only to the data file without the DV
```

Comment:
Love the context comments.  This makes it very easy to follow what's going on in the test.